Improved SVG support

Aiving opened the pull request:
This PR is intended to improve the handling of SVGs in `skia-safe`, and to add even more SVG rendering capabilities from Skia. However, it is still incomplete and requires some discussion. A list of things that have not yet been ported will be announced at the end.

As for the things that require discussion, I'll tell you about them in order.

## A new crate containing a procedural macro

`skia-macros` is a somewhat forced solution. Personally, I don't like having to go to such lengths, but I had a choice: either write a mountain of identical, similar code for the attributes of each of the possible elements, or use macros. I still chose the second option, however it was originally a simpler macro written with `macro_rules!`. Its problem was the parameters. In some cases, such as with `get_opacity`, it was necessary that the return type not be a reference, since `f32` implements `Copy`. In the same case with `get_opacity`, you also had to check if it was specified, since simply checking for a reference to `null` was useless in the case of numbers. And here `macro_rules!` became not useless, but much more unpleasant to use.

However, adding an entire procedural macro for the sake of just one of the modules seems a bit overkill. My current solution is to make it `optional` and put it as `dep:skia-macos` under the `svg` feature.  

## ~~`SvgNode`, `SvgSpecNode<T>`, `Tagged` and `AsTag`~~

### Solution: removing `SvgSpecNode<T>` and `AsTag`, refactoring `SvgNode` and using `Deref`/`DerefMut`

~~Obviously, `SvgNode` and `SvgSpecNode<T>` are two almost identical in meaning and code, structures. However, the problem is that `Dom::get_root` should always return `Svg`, with the ability to access more basic attributes like `fill` or `opacity`. Using `SvgNode` here is **possible**, but **desirable**, as it would add unnecessary hassle for library users (having to check every time whether the `SvgNode::attributes` value returned is `NodeAttributes::Svg`, even though it **always** is).~~

~~Because of this, `SvgSpecNode<T>` appeared, and with it two new traits - `Tagged` to find all structures that have `Tag` and `AsTag` to implement certain functions for them and to limit the number of types that can be used as a generic parameter in `SvgSpecNode<T>`.~~

~~In general, I don't have any problems with traits, but I have problems with having two identical structures. At least the fact that the code for `SvgSpecNode<T>` looks rather dirty.~~

## Acknowledgements

I am truly grateful to @marc2332 for his initial contribution, as it was his PR (unfortunately closed) that helped me decide to create my own.

### Progress on implementing wrappers for types from the SVG module

- [x] Circle
- [x] Clip Path
- [x] Defs
- [x] Ellipse
- [x] Fe Blend
- [x] Fe Color Matrix
- [x] Fe Component Transfer
- [x] Fe Composite
- [x] Fe Diffuse Lighting
- [x] Fe Displacement Map
- [x] Fe Distant Light
- [x] Fe Flood
- [x] Fe Func A
- [x] Fe Func R
- [x] Fe Func G
- [x] Fe Func B
- [x] Fe Gaussian Blur
- [x] Fe Image
- [x] Fe Merge
- [x] Fe Merge Node
- [x] Fe Morphology
- [x] Fe Offset
- [x] Fe Point Light
- [x] Fe Specular Lighting
- [x] Fe Spot Light
- [x] Fe Turbulence
- [x] Filter
- [x] G
- [x] Image
- [x] Line
- [x] Linear Gradient
- [x] Mask
- [x] Path
- [x] Pattern
- [x] Polygon
- [x] Polyline
- [x] Radial Gradient
- [x] Rect
- [x] Stop
- [x] Svg
- [x] Text
- [x] Text Literal
- [x] Text Path
- [x] TSpan
- [x] Use

Aiving commented:
@pragmatrix Do you have any suggestions as to why the structure for `SkSVGProperty` is not created by `bindgen`? I thought it was because of the use of C++ templates, but it looks like `bindgen` handles them just fine.

I'm also wondering if it's possible to fix the error in layout tests without adding the type to the list of opaque types?

pragmatrix commented:
Regarding SkSVGProperty, looks like non-type template parameters aren't supported yet: https://github.com/rust-lang/rust-bindgen/issues/362

pragmatrix commented:
> I'm also wondering if it's possible to fix the error in layout tests without adding the type to the list of opaque types?

Not that I am aware of. But usually it's not the problem with the type you need, but with a type it consists of. So when multiple layout errors appear, it's often good enough to make the smallest / simplest type opaque.

Aiving commented:
> Regarding SkSVGProperty, looks like non-type template parameters aren't supported yet: [rust-lang/rust-bindgen#362](https://github.com/rust-lang/rust-bindgen/issues/362)

To be honest, it's a pity, because the solution of the mentioned issue has been delayed until const generics are stabilised. And judging by their repository, they will be stabilised obviously not soon.

Aiving commented:
@pragmatrix BTW, do you have any ideas regarding the `SvgSpecNode<T>` situation?

pragmatrix commented:
I am currently sick and have trouble concentrating. I will try to get into the details as soon I am better.

pragmatrix commented:
> Dom::get_root should always return Svg, with the ability to access more basic attributes like fill or opacity.

I am not sure, but I think that this upcasting problem can be solved using the the `Deref()` / `DerefMut()` trait directly on the `RCHandle<SkSVGSVG>`. `&RCHandle<N>` are transmutable to any other `&RCHandle<NB>` where NB is a C++ base class of N.

Aiving commented:
> > Dom::get_root should always return Svg, with the ability to access more basic attributes like fill or opacity.
> 
> I am not sure, but I think that this upcasting problem can be solved using the the `Deref()` / `DerefMut()` trait directly on the `RCHandle<SkSVGSVG>`. `&RCHandle<N>` are transmutable to any other `&RCHandle<NB>` where NB is a C++ base class of N.

Thanks for the advice! To be honest, I only recently learnt that `Deref`/`DerefMut` can be used to simulate inheritance similar to that in C++
I did some refactoring by removing `SvgSpecNode<T>` and replacing it with a new `SvgNode<T>` (with improved implementation of the `Debug` trait), which is also used as a replacement for `RCHandle<T>` for `Svg*` types. Also added some more wrappers for types from the `svg` module.

Aiving commented:
@pragmatrix This PR is almost at the final stage, although some things still need to be finalized (I still haven't linked `skia_macros` to the `svg` feature and I think some structures should be moved to the `types` module). So I'd like to get your opinion on it. Also, as you can see, I decided to change the system a bit again, replacing `SvgNode<T>` with a new structure `Inherits<N, B>` which allows to create a more flexible pseudo-inheritance.

pragmatrix commented:
I've a comment related to Inherit first.

I think you can base everything on `RCHandle<T>` and implement a trait that specifies the through an associated type and then implement Deref and DerefMut on top of all types that implement that trait.

I've experimented a little, and something like this should work:

```
pub trait HasBase {
    type Base;
}

impl<NT> Deref for RCHandle<NT>
where
    NT: NativeRefCounted,
    RCHandle<NT>: HasBase,
{
    type Target = <RCHandle<NT> as HasBase>::Base;

    fn deref(&self) -> &Self::Target {
        todo!()
    }
}
```
The implementation is a simple transmute on the reference, because underlying they are the same thing.

But this introduces a safety issue with `DerefMut()`, because one could just overwrite the RCHandle<Base> with a Base that is not compatible with the derived C++ class. Two ideas here: Methods could either check the tag before each call to C++ or implement every method (even the mutating ones) so that they work on shared references.

May be you've already considered that and therefore moved to storing the base ptr separately even though they are the same thing?

Aiving commented:
> May be you've already considered that and therefore moved to storing the base ptr separately even though they are the same thing?

Not really. I just thought that from an optimisation point of view, having a structure that already has a `base` property would be much better for implementing `Deref`/`DerefMut` than a constant cast to the right type. Although I am very confused by the fact that `from_unshared_ptr` is called not only for `current` type, but also for `base` type.

> I've experimented a little, and something like this should work:
> 
> ```
> pub trait HasBase {
>     type Base;
> }
> 
> impl<NT> Deref for RCHandle<NT>
> where
>     NT: NativeRefCounted,
>     RCHandle<NT>: HasBase,
> {
>     type Target = <RCHandle<NT> as HasBase>::Base;
> 
>     fn deref(&self) -> &Self::Target {
>         todo!()
>     }
> }
> ```
> 
> The implementation is a simple transmute on the reference, because underlying they are the same thing.

It is not quite clear to me how it should work. The `RCHandle` has no methods to "create" a reference to it (like `from_native_ref`), and a cast like `self as RCHandle::<NT::Base>` triggers an `infinite loop` warning

pragmatrix commented:
I've no idea why the windows CI builds fail. I've tried locally on one of my machines, updated VS, Rust, removed `Cargo.lock`, did a `cargo clean`, and there it builds just fine.

Aiving commented:
> I've no idea why the windows CI builds fail. I've tried locally on one of my machines, updated VS, Rust, removed `Cargo.lock`, did a `cargo clean`, and there it builds just fine.

Yeah, strange behaviour.

marc2332 commented:
> I've no idea why the windows CI builds fail. I've tried locally on one of my machines, updated VS, Rust, removed `Cargo.lock`, did a `cargo clean`, and there it builds just fine.

CI cache maybe?

pragmatrix commented:
The errors are caused by LLVM / clang 19.1. 18.1 works, but interestingly 19.1 succeeds in master with the svg feature enabled, so this PR causes clang 19 to behave in a different way for some reason. I'll try to figure this out independently of the progress here.

pragmatrix reviewed (changes requested):
So this looks fairly complete now. I'd suggest two more changes:

- Is it reasonable to remove the `Svg` prefix? `SvgDisplay` and `SvgColorSpace`   clash with other types, but the types are under the module `svg` and can be qualified if needed. What do you think @Aiving and @marc2332, @triniwiz ?
- As discussed before, I'd suggest to remove the `Inherited<>` struct completely and base everything on RcHandle<T>. This would remove a lot of duplicated code (e.g. all the `from_ptr`, `from_unshared_ptr`). @Aiving references can be transformed using `prelude::transform_ref` for example to return a baseclass `RCHandle<>` from a Deref implementation. And then - to be sound on the Rust side - remove `DerefMut` and allow all modifications on shared references.

triniwiz commented:
@pragmatrix Sounds 💯

Aiving commented:
> * allow all modifications on shared references.

@pragmatrix Can you give me an example? I'm not sure I got this right.

Aiving commented on skia-bindings/src/svg.cpp:
```diff
@@ -75,3 +114,371 @@ extern "C" bool C_SkSVGDOM_unique(const SkSVGDOM* self) {
+SVG_ATTRIBUTE(SkSVGUse, Y   , SkSVGLength);
+SVG_ATTRIBUTE(SkSVGUse, Href, SkSVGIRI   );
+
+extern "C" void C_SkSVGIRI_new(SkSVGIRI* uninitialized) {
```
resolved in 98bfbdc098c70505c28126bf6e4f469cf9441fd0

Aiving commented on skia-bindings/src/svg.cpp:
```diff
@@ -75,3 +114,371 @@ extern "C" bool C_SkSVGDOM_unique(const SkSVGDOM* self) {
+    return static_cast<const SkSVGContainerAccessor&>(self).children();
+}
+
+extern "C" void C_SkSVGTransformableNode_setTransform(SkSVGTransformableNode* self, const SkMatrix value) {
```
resolved in 98bfbdc098c70505c28126bf6e4f469cf9441fd0

pragmatrix commented:
@Aiving Whenever a function is implemented that modifies some attribute and therefore takes a `&mut self`, change this to `&self` and then use `native_mut_force()` before calling the C function. I suppose this should be reasonably simple to change because of the use of macros.

Aiving commented:
> @Aiving Whenever a function is implemented that modifies some attribute and therefore takes a `&mut self`, change this to `&self` and then use `native_mut_force()` before calling the C function. I suppose this should be reasonably simple to change because of the use of macros.

Wouldn't that violate the Rust rule about only being able to have one mutable reference at a time? I'm not saying I'm against it, but I'm wondering if it would cause any problems in the future

pragmatrix commented:
> Wouldn't that violate the Rust rule about only being able to have one mutable reference at a time?

It may, but only indirectly, if a `&mut` reference would be returned through `&self` invocations for example. Effectively though, there is only ever a single temporary `&mut T` in play as long all mutations happen in C++, and all the `RcHandle<T>` are cloned / recreated when T gets reused at other places. And also, `RcHandle<T>`s are not safe to be sent to other threads when their reference counter is >1.

BTW: All Canvas functions are taking `&self`, because `Surface` once returned a `&mut Canvas`: #814

We can leave the `DerefMut` if you want for now. I am not strict here regarding potential misuses of the API that lead to crashes when they were crafted intentionally. Who would dare to overwrite the `RcHandle<T>` returned by a `DerefMut`?

Aiving commented:
> > Wouldn't that violate the Rust rule about only being able to have one mutable reference at a time?
> 
> It may, but only indirectly, if a `&mut` reference would be returned through `&self` invocations for example. Effectively though, there is only ever a single temporary `&mut T` in play as long all mutations happen in C++, and all the `RcHandle<T>` are cloned / recreated when T gets reused at other places. And also, `RcHandle<T>`s are not safe to be sent to other threads ~when their reference counter is >1~.
> 
> BTW: All Canvas functions are taking `&self`, because `Surface` once returned a `&mut Canvas`: #814
> 
> We can leave the `DerefMut` if you want for now. I am not strict here regarding potential misuses of the API that lead to crashes when they were crafted intentionally. Who would dare to overwrite the `RcHandle<T>` returned by a `DerefMut`?

I think I'm done!

pragmatrix commented:
Great work, add yourself to the README.md under notable contributions if you want to.

Aiving commented:
> Great work, add yourself to the README.md under notable contributions if you want to.

done

marc2332 commented:
Amazing work @Aiving, I'll try this for my pr in freya and provide feedback

marc2332 commented:
I am not sure what is happening but I can't get skia to compile. Any clue?
<details>

<summary>click here to see the logs</summary>

```
   Compiling skia-bindings v0.78.0 (C:\Users\mespi\rust-skia-1030\skia-bindings)
error: failed to run custom build command for `skia-bindings v0.78.0 (C:\Users\mespi\rust-skia-1030\skia-bindings)`

Caused by:
  process didn't exit successfully: `C:\Users\mespi\rust-skia-1030\target\debug\build\skia-bindings-d2459326d46671ef\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-env-changed=DOCS_RS
  cargo:rerun-if-env-changed=SKIA_DEBUG
  HOST: x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=SKIA_SOURCE_DIR
  cargo:rerun-if-env-changed=FORCE_SKIA_BUILD
  cargo:rerun-if-env-changed=FORCE_SKIA_BINARIES_DOWNLOAD
  STARTING A FULL BUILD
  cargo:rerun-if-env-changed=CLANGCC
  cargo:rerun-if-env-changed=CC
  cargo:rerun-if-env-changed=CLANGCXX
  cargo:rerun-if-env-changed=CXX
  HOST: x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=OPT_LEVEL
  cargo:rerun-if-env-changed=SKIA_USE_SYSTEM_LIBRARIES
  HOST: x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=SDKTARGETSYSROOT
  cargo:rerun-if-env-changed=SDKROOT
  HOST: x86_64-pc-windows-msvc
  HOST: x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=VCINSTALLDIR
  cargo:rerun-if-env-changed=LLVM_HOME
  cargo:rerun-if-env-changed=USERPROFILE
  cargo:rerun-if-env-changed=SKIA_NINJA_COMMAND
  cargo:rerun-if-env-changed=SKIA_GN_COMMAND
  Probing 'python'
  Python 3 found: "python"
  Synchronizing Skia dependencies
  Skipping "bin".
  ...ird_party/externals/libjpeg-turbo @ ccfbe1c82a3b6dbe8647ceb36a3f9ee711fba3cf
  skia\third_party/externals/wuffs     @ e3f919ccfe3ef542cfc983a82146070258fb57f8
  skia\third_party/externals/libwebp   @ 845d5476a866141ba35ac133f856fa62f0b7445f
  ...rd_party/externals/d3d12allocator @ 169895d529dfce00390a20e69c2f516066fe7a3b
  skia\third_party/externals/zlib      @ 646b7f569718921d7d4b5b8e22572ff6c76f2596
  skia\third_party/externals/expat     @ 441f98d02deafd9b090aea568282b28f66a50e36
  skia\third_party/externals/harfbuzz  @ b74a7ecc93e283d059df51ee4f46961a782bcdb8
  ...y/externals/vulkanmemoryallocator @ a6bfc237255a6bac1513f7c1ebde6d8aed6b5191
  skia\third_party/externals/libpng    @ ed217e3e601d8e462f7fd1e04bed43ac42212429
  skia\third_party/externals/brotli    @ 6d03dfbedda1615c4cba1211f8d81735575209c8
  skia\third_party/externals/freetype  @ 73720c7c9958e87b3d134a7574d1720ad2d24442
  ...third_party/externals/spirv-cross @ b8fcf307f1f347089e3c46eb4451d27f32ebc8d3

  --- stderr
  Checking for "C:\\Program Files\\LLVM\\bin\\clang-cl.exe"
  error: The following untracked working tree files would be overwritten by checkout:
        APIChangeReport.html
        BUILD.gn
        DIR_METADATA
        LICENSE
        OWNERS
        README.chromium
        README.fuchsia
        android/icudtl.dat
        cast/brkitr.patch
        cast/icudtl.dat
        cast/patch_locale.sh
        chromeos/icudtl.dat
        chromeos/icudtl.dat.hash
        codereview.settings
        common/icudtb.dat
        common/icudtl.dat
        config.gni
        filters/android-extra-removed-resources.txt
        filters/android.json
        filters/cast.json
        filters/chromeos.json
        filters/common.json
        filters/flutter.json
        filters/flutter_desktop.json
        filters/ios.json
        flutter/README.md
        flutter/brkitr.patch
        flutter/icudtl.dat
        flutter/patch_brkitr.sh
        flutter_desktop/icudtl.dat
        fuzzers/BUILD.gn
        fuzzers/OWNERS
        fuzzers/fuzzer_utils.h
        fuzzers/icu_appendable_fuzzer.cc
        fuzzers/icu_break_iterator_fuzzer.cc
        fuzzers/icu_break_iterator_utf32_fuzzer.cc
        fuzzers/icu_converter_fuzzer.cc
        fuzzers/icu_number_format_fuzzer.cc
        fuzzers/icu_regex.dict
        fuzzers/icu_to_case_fuzzer.cc
        fuzzers/icu_ucasemap_fuzzer.cc
        fuzzers/icu_unicode_string_codepage_create_fuzzer.cc
        fuzzers/icu_uregex_open_fuzzer.cc
        icu.gyp
        icu.gypi
        icu.isolate
        icu4c.css
        icu_nacl.gyp
        ios/icudtl.dat
        license.html
        patches/ardatepattern.patch
        patches/atomic_template_instantiation.patch
        patches/cjdict.patch
        patches/configure.patch
        patches/data_symb.patch
        patches/fuchsia.patch
        patches/gb_table.patch
        patches/iso2022jp.patch
        patches/khmer-dictbe.patch
        patches/locale1.patch
        patches/locale_google.patch
        patches/name_5_langs.patch
        patches/restrace.patch
        patches/revert_realpath.patch
        patches/wordbrk.patch
        patches/wpo.patch
        readme.html
        scripts/LICENSE
        scripts/accept_lang.list
        scripts/asm_to_inline_asm.py
        scripts/big5_gen.sh
        scripts/chrome_ui_languages.list
        scripts/clean_up_data_source.sh
        scripts/copy_data.sh
        scripts/currencies.list
        scripts/data_common.sh
        scripts/data_files_to_preserve.txt
        scripts/diff_data.sh
        scripts/diff_data_all.sh
        scripts/diff_res.sh
        scripts/eucjp_gen.sh
        scripts/euckr_gen.sh
        scripts/ibm866_gen.sh
        scripts/icualign.py
        scripts/icuhash.py
        scripts/make_data_all.sh
        scripts/make_data_assembly.py
        scripts/make_data_cpp.py
        scripts/make_mac_assembly.sh
        scripts/single_byte_gen.sh
        scripts/sjis_gen.sh
        scripts/update.sh
        scripts/update_tz.sh
        shim_headers.gypi
        source/.clang-format
        source/Doxyfile.in
        source/Makefile.in
        source/acinclude.m4
        source/aclocal.m4
        source/common/BUILD.bazel
        source/common/Makefile.in
        source/common/appendable.cpp
        source/common/bmpset.cpp
        source/common/bmpset.h
        source/common/brkeng.cpp
        source/common/brkeng.h
        source/common/brkiter.cpp
        source/common/bytesinkutil.cpp
        source/common/bytesinkutil.h
        source/common/bytestream.cpp
        source/common/bytestrie.cpp
        source/common/bytestriebuilder.cpp
        source/common/bytestrieiterator.cpp
        source/common/caniter.cpp
        source/common/capi_helper.h
        source/common/characterproperties.cpp
        source/common/chariter.cpp
        source/common/charstr.cpp
        source/common/charstr.h
        source/common/charstrmap.h
        source/common/cmemory.cpp
        source/common/cmemory.h
        source/common/common.rc
        source/common/cpputils.h
        source/common/cstr.cpp
        source/common/cstr.h
        source/common/cstring.cpp
        source/common/cstring.h
        source/common/cwchar.cpp
        source/common/cwchar.h
        source/common/dictbe.cpp
        source/common/dictbe.h
        source/common/dictionarydata.cpp
        source/common/dictionarydata.h
        source/common/dtintrv.cpp
        source/common/edits.cpp
        source/common/emojiprops.cpp
        source/common/emojiprops.h
        source/common/errorcode.cpp
        source/common/filteredbrk.cpp
        source/common/filterednormalizer2.cpp
        source/common/hash.h
        source/common/icudataver.cpp
        source/common/icuplug.cpp
        source/common/icuplugimp.h
        source/common/loadednormalizer2impl.cpp
        source/common/localebuilder.cpp
        source/common/localefallback_data.h
        source/common/localematcher.cpp
        source/common/localeprioritylist.cpp
        source/common/localeprioritylist.h
        source/common/localsvc.h
        source/common/locavailable.cpp
        source/common/locbased.cpp
        source/common/locbased.h
        source/common/locdispnames.cpp
        source/common/locdistance.cpp
        source/c
  Aborting
  Traceback (most recent call last):
    File "C:\Users\mespi\rust-skia-1030\skia-bindings\skia\tools\git-sync-deps", line 281, in <module>
      exit(main(sys.argv[1:]))
    File "C:\Users\mespi\rust-skia-1030\skia-bindings\skia\tools\git-sync-deps", line 269, in main
      git_sync_deps(deps_file_path, argv, shallow, verbose)
    File "C:\Users\mespi\rust-skia-1030\skia-bindings\skia\tools\git-sync-deps", line 239, in git_sync_deps
      multithread(git_checkout_to_directory, list_of_arg_lists)
    File "C:\Users\mespi\rust-skia-1030\skia-bindings\skia\tools\git-sync-deps", line 256, in multithread
      raise Exception("Thread failure detected")
  Exception: Thread failure detected
  thread 'main' panicked at skia-bindings\build_support\skia\config.rs:282:9:
  `skia/tools/git-sync-deps` failed
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

</details>

Aiving commented:
> I am not sure what is happening but I can't get skia to compile. Any clue?
> 
> Notes, I even made a fresh clone, but no luck.

It's pretty weird. It doesn't even seem to be an error that appears in checks for windows. The files look like `icu`, however I can't even imagine what causes this to happen. @pragmatrix do you have any ideas?

pragmatrix commented:
Never had this before. This is related to git-sync-deps, which attempts to check out the icu dependency into the directory skia-bindings/skia/third_party/externals/icu. If something happens there, I usually delete everything under externals and try again. It looks to me that may be some git settings are interfering (global CRLF settings may be?), or some other programs work there in parallel? @marc2332 Try to go into the directory and do a git status / diff, may be that sheds some light on the issue.

marc2332 commented:
> Never had this before. This is related to git-sync-deps, which attempts to check out the icu dependency into the directory skia-bindings/skia/third_party/externals/icu. If something happens there, I usually delete everything under externals and try again. It looks to me that may be some git settings are interfering (global CRLF settings may be?), or some other programs work there in parallel? @marc2332 Try to go into the directory and do a git status / diff, may be that sheds some light on the issue.

Indeed, somehow there are changes in that folder, I don't understand how or why as  it was a fresh clone. seems to work now, I am gonna try it on freya

marc2332 commented on skia-macros/src/lib.rs:
```diff
@@ -0,0 +1,257 @@
+                quote! {
+                    #getter
+
+                    pub fn #set_name(&self, #setter_name: #ty) {
```
For the sake of making things more rusty I feel like this would be nice, but in any case, LGTM!
```suggestion
                    pub fn #set_name(&mut self, #setter_name: #ty) {
```

marc2332 reviewed (approved):
Just had to adapt to the API changes, but I am still able to force the root element to have the size I wanted ✨ 
![image](https://github.com/user-attachments/assets/cc750a80-a80f-4125-8121-e4d6a8d1b8ab)

pragmatrix commented:
> For the sake of making things more rusty I feel like this would be nice, but in any case, LGTM!

@marc2332 I am reconsidering this decision to allow modification on `&self` to avoid the implementation of DerefMut. Perhaps we should first make it idiomatic and then make it super-safe. We can always get on `&self` later without causing a lot of trouble, but not the other way around.

I've fixed the merge conflict, and with #1043 landed on master, this should build on windows, too.

@Aiving I assume this is ready from your side. However, there are some minor issue I need to address before a first release, some related to naming (I want to get rid of Fe and the remaining Svg prefixes, also there are some enum types exposed that are prefixed with kSomething) , and some potential memory leaks with native transmutable types (SkString). But I've decided to merge this PR first and then follow up on in a separate PR soon if this is fine with you.

Aiving commented:
> @Aiving I assume this is ready from your side. However, there are some minor issue I need to address before a first release, some related to naming (I want to get rid of Fe and the remaining Svg prefixes, also there are some enum types exposed that are prefixed with kSomething) , and some potential memory leaks with native transmutable types (SkString). But I've decided to merge this PR first and then follow up on in a separate PR soon if this is fine with you.

Ofc! Just remembered in the process of reading this post that I forgot to finish renaming a couple enums.
